### PR TITLE
issue #911 Sign new message if the message user replying to is signed

### PR DESCRIPTION
**Pull request by michael-volynets**
Issue #911 

Time Spent: 1h

When I was working on this issue I found another issue, when we saving signed draft Flowcrypt saves it as encrypted msg (not a signed msg). I was trying to fix it, and I fixed but the duration of draft saving (if signed) increased in two or more times so I removed the changes from this PR.
Maybe we need to create another issue to make an ability to save signed drafts, what do you think?

### CODE REVIEW

**Comment by tomholub** [comment]
Good observation (on the speed). Signing is an expensive operation (similar to decrypting). I'm thinking we will not support signed drafts at all for that reason.

RSA signing, decrypting - expensive
RSA encrypting, verifying - (relatively) cheap

File: extension/js/common/composer.ts [hunk 1]
```diff
@@ -1418,12 +1424,12 @@ export class Composer {
         progressCb ? (progress: number) => progressCb(progress * 0.6) : undefined);
       const message = await Mime.process(Buf.fromBase64UrlStr(raw!));
       const readableBlocks = message.blocks
-        .filter(b => b.type === 'encryptedMsg' || b.type === 'plainText' || b.type === 'plainHtml');
-      const encryptedCount = readableBlocks.filter(b => b.type === 'encryptedMsg').length;
+        .filter(b => ['encryptedMsg', 'plainText', 'plainHtml', 'signedMsg'].includes(b.type));
+      const encryptedCount = readableBlocks.filter(b => ['encryptedMsg', 'signedMsg'].includes(b.type)).length;
```
Comment (tomholub) [inline extension/js/common/composer.ts]: Maybe `encryptedCount` could be renamed to `pgpBlockCount` (purpose has changed)

Comment (michael-volynets) [reply extension/js/common/composer.ts]: Yes, agree, 
from now it stores count for both encrypted and signed blocks

**Comment by michael-volynets** [comment]
Okay, thanks,
I forgot about the test,
And maybe, you have some cases when I can be 100% sure that I need to add the tests for the issue? Because sometimes I just don't know whether I need to add tests or not

**Comment by tomholub** [comment]
Here is my rule of thumb when I'm not sure if I should be adding test. I ask myself:

1) Is it manageable to manually test this feature after other code changes to make sure it still works?
2) If it fails in the future, would it be easy to discover?
3) If it fails, is it a big deal?

If answer to 1 or 2 is no, then I probably want to write a test.

But then if answer to 3 is also no, maybe it's ok to skip a test.

In this case, I would not want to have to manually test this after changes, so a test may be appropriate. But then, if it fails, it's not a big deal (in this case). So the answer is subjective.

The reason I thought to make a test in this case is that I want to re-work the "sign and send" UX in the future (to be a button with a down chevron for message type: sign/encrypt/sign+encrypt/plain). I'm afraid this functionality will be easy to overlook when we rework it, so I would like to add a test for that reason.

Overall, now that we use mocks for testing, reliability of the tests should generally not be a problem. So more tests are a good thing, as long as they test something useful.

**Comment by tomholub** [comment]
I just thought, for composer changes specifically, I would lean on the side of adding tests, because composer is quite complex yet quite easy to test independently.

**Comment by michael-volynets** [comment]
Thanks for the explanation, I'll try to follow these rules.

I also have a question, is there a way to get HTMLElement in the tests?
I mean, to test whether the reply box is signed I want to check if there is '.active' class on sign icon:
image

But I can't access this sign icon,
I tried using `.target.$('.icon.action_sign'))` and it returns  me `ElementHandle<Element>` object but I can't access `class` attribute through that object.

Maybe you have an idea of how to access it?

**Comment by tomholub** [comment]
Two approaches

One approach is to try to figure out the class from found element. You could try something like
```js
let signIcon = await page.waitAny('@action-toggle-sign-message');
let classProperty = signIcon.getProperty("class");
let classJson = classProperty.jsonValue();  // evaluate the classJson object, see if you can get the class from there
```
I'm just guessing from https://github.com/GoogleChrome/puppeteer/blob/v1.19.0/docs/api.md#class-elementhandle - if you will take this approach, it would make sense to add it as a static convenience method, maybe on `PageController`.

The other option is to query it including the intended class eg `.icon.action_sign.whatever_intended_class`

It would be better to use `@selectors` that we typically have for testing. You could try to edit this function:
```js
  protected selector = (customSelLanguageQuery: string): string => { // supply browser selector, xpath, @test-id or @test-id(contains this text)
    let m;
    if (this.isXpath(customSelLanguageQuery)) {
      return customSelLanguageQuery;
    } else if (m = customSelLanguageQuery.match(/^@([a-z0-9\-]+)$/)) { // tslint:disable-line:no-conditional-assignment
      return `[data-test="${m[1]}"]`;
    } else if (m = customSelLanguageQuery.match(/^@([a-z0-9\-]+)\(([^()]*)\)$/)) { // tslint:disable-line:no-conditional-assignment
      return `//*[@data-test='${m[1]}' and contains(text(),'${m[2]}')]`;
    } else {
      return customSelLanguageQuery;
    }
  }
```
Maybe add something like:
```js
    } else if (m = customSelLanguageQuery.match(/^@([a-z0-9\-]+)\.[a-z\-_]+$/)) { // tslint:disable-line:no-conditional-assignment
      return `[data-test="${m[1]}"].${m[2]}`;
```
Now our selectors will be a bit smarter and you can select a specific class with it:
```js
let signIcon = await page.waitAll('@action-toggle-sign-message.whatever_intended_class');
```
Which should allow you to achieve what you want.

I haven't tested any of this code.

**Comment by michael-volynets** [comment]
Thanks a lot,

I used the first approach in the PR

**Comment by tomholub** [comment]
@michael-volynets I can merge this when the conflicts are resolved

**Comment by michael-volynets** [comment]
Resolved

**Pull request merged**